File: PacketMetrics/MATIEAllMethods.r
#!/usr/bin/env Rscript
#-----------------------------------------------------
#-----------------------------------------------------
#--                Function Name: TDEV              --
#--                Name: Time Deviation             --
#--           Input: nTo - position in list         --
#--                  N   - number of samples        -- 
#--                  x   - vector of samples        --
#--           Output : time deviation               --
#-----------------------------------------------------
#-----------------------------------------------------
MATIEAllMethods <- function(To,n, N,x){
#	To <- 0.1 #time between samples
#	n <- nTo / To #number of samples to current point
	window <- 5 # Set window Size
	windowSide <- (window - 1) / 2 # Set the length of Side of window
	outerStep <- 0 
	#MATIE, MAFE, minMATIE, MAFE
	result <- matrix(0,4)
	interimResult <- matrix(0,(N - 2*n + 10),4) #fudged because of the 4 truncated rows. will fix what N is. 
	interimStep <- matrix(0,2) #only need MATIE and minMATIE
	for (i in 1:(N-2*n + 1)){
		interimStep <- c(0,0,0,0)
		for (j in i + windowSide:(n+i-1) - windowSide){
			interimStep[0] <- interimStep[0] + mean(x[i+n - windowSide: i + n + windowSide]) +  mean(x[i - windowSide : i + windowSide])
			interimStep[1] <- interimStep[1] + min(x[i + n - windowSide : i + n + windowSide]) + min(x[i - windowSide : i + windowSide])
		}
		for (k in 1:2){
		
			interimStep[k] <- abs(interimStep[k]) / n
		}
		interimResult[i,1] <- interimStep[1]
		interimResult[i,2] <- interimStep[1] / (n * To)
		interimResult[i,3] <- interimStep[2]
		interimResult[i,4] <- interimStep[2] / (n * To)
	}
	result[1] <- max(interimResult[1])		
	result[2] <- max(interimResult[2])		
	result[3] <- max(interimResult[3])		
	result[4] <- max(interimResult[4])		
	return(result)
}

File: PacketMetrics/MATIEAllMethods.c
//-----------------------------------------------------
//-----------------------------------------------------
//--                Function Name: TDEV              --
//--                Name: Time Deviation             --
//--           Input: nTo - position in list         --
//--                  N   - number of samples        -- 
//--                  x   - vector of samples        --
//--           Output : time deviation               --
//-----------------------------------------------------/
//-----------------------------------------------------
#include <math.h>
#include <stdio.h>
#include <R.h>

void MATIEAllMethods(double *To, int* n, int* N ,double * x,double * tempResultMATIE,double * tempResultMAFE, double * tempResultMinMATIE, double * tempResultMinMAFE){
	int window = 5; // Set window Size
	int windowSide =  (window - 1) / 2; // Set the length of Side of window
	double minimum[2] = {10000,10000};	
	double interimStep[2][*N - (2 * (*n)) + 10];
	double average[2] = {0.0,0.0};
	for (int i=0; i <= (*N -(2 * (*n)))  + 1 -1; i++){
 		for (int j=i; j < *n + i - 1 - 1 ; j++) { // - 1 as normal. - 1 for index difference
			for (int k=0;k < window ;k++){
				average[0] = average[0] + x[j + (*n) - windowSide + k - 1];
				average[1] = average[1] + x[j - windowSide + k - 1];
				if (x[j + (*n) - windowSide + k - 1] < minimum[0]) minimum[0] = x[j + (*n) - windowSide + k - 1];
				if (x[j- windowSide + k - 1] < minimum[1] ) minimum[1] = x[j - windowSide + k - 1];
			
			}
		}
		for (int j = 0; j < 2; j++) {
			average[j] = average[j] / window;
		}
		
		interimStep[0][i] = (average[0] - average[1]) < 0 ? -((average[0] - average[1]) / *n) : ((average[0] - average[1]) / *n);
		interimStep[1][i] = (minimum[0] - minimum[1]) < 0 ? -(minimum[0] - minimum[1]) / *n : (minimum[0] - minimum[1]) / *n;
		average[0] = average[1] = 0;
		minimum[0] = minimum[1] = 10000.00;
		if (interimStep[0][i] > *tempResultMATIE) *tempResultMATIE = interimStep[0][i];
		if (interimStep[1][i] > *tempResultMinMATIE) *tempResultMinMATIE = interimStep[1][i];
		*tempResultMAFE = (double) *tempResultMATIE / (*n * (*To));
		*tempResultMinMAFE = (double) *tempResultMinMATIE / (*n * (*To));
		
	}

}

